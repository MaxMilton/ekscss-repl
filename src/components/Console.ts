/* eslint-disable @typescript-eslint/no-explicit-any */

import './Console.xcss';

import {
  append, create, html, S1Node,
} from 'stage1';
import { refs } from '../service';

export type ConsoleComponent = S1Node &
HTMLDivElement & {
  log(...args: any[]): void;
  warn(...args: any[]): void;
  error(...args: any[]): void;
  clear(): void;
};

type RefNodes = {
  con: HTMLDivElement;
};

const view = html`
  <div id="con">
    <h2>Console Output</h2>
    <div class="console code-block code" #con></div>
  </div>
`;

type ConsoleMethodName = 'log' | 'warn' | 'error';

export function Console(): ConsoleComponent {
  const root = view as ConsoleComponent;
  const { con } = view.collect<RefNodes>(root);

  const print = (method: ConsoleMethodName, color = '') => (...args: unknown[]) => {
    const line = create('div');
    line.className = color;
    line.textContent = args.toString();
    append(line, con);
    con.scrollTo(0, con.scrollHeight);

    // eslint-disable-next-line no-console
    console[method](...args);
  };

  root.log = print('log');
  root.warn = print('warn', 'gold4');
  root.error = print('error', 'red5');

  root.clear = () => {
    con.textContent = '';
  };

  append(con, root);
  refs.console = root;

  return root;
}
