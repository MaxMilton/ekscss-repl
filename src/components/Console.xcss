#console {
  max-height: 240px;
  overflow: auto;
  contain: layout style paint;
}
