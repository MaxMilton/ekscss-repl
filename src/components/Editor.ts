import { h, S1Node } from 'stage1';
import './Editor.xcss';

export type EditorComponent = S1Node &
HTMLOListElement & {
  setContent(code: string): void;
  getContent(): string;
};

const view = h`
  <ol contenteditable=true class="editor textarea w100 mt0 code" rows=10 spellcheck=false></ol>
`;

export function Editor(): EditorComponent {
  const root = view.cloneNode(true) as EditorComponent;

  document.execCommand('defaultParagraphSeparator', false, 'li');

  // convert rich content to plain text when pasting
  root.onpaste = (event) => {
    event.preventDefault();
    const text = event.clipboardData?.getData('text/plain');
    document.execCommand('insertText', false, text);
  };

  root.setContent = (code) => {
    // TODO: Using innerHTML could be a security issue so consider a refactor
    // to use textContent/innerText and DOM methods
    root.innerHTML = code
      .split('\n')
      .map((line) => `<li>${line || '<br>'}</li>`)
      .join('');
  };

  // use innerText instead of textContent here so we also get \n etc.
  // eslint-disable-next-line unicorn/prefer-dom-node-text-content
  root.getContent = () => root.innerText;

  return root;
}
