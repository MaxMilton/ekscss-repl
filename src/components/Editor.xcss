.editor {
  // FIXME: This fixes the .error-line overflow background not full-width but
  // doesn't feel like a great solution
  // display: grid;

  max-width: 100%;
  height: 100%;
  overflow-x: auto;
  padding: 0;
  text-indent: 1em;
  white-space: pre;
  list-style-position: inside;
  list-style-type: decimal-leading-zero;
  list-style: ;
  contain: layout paint style;

  & > li {
    // FIXME: Doesn't work
    // highlight current line selection
    // &:focus-within {
    &:focus,
    &:hover {
      background-color: ${x.color.light4};
    }

    &::marker {
      color: ${x.color.gray2};
    }
  }
}

.editor-wrap {
  white-space: normal;
  word-break: break-all;
}
