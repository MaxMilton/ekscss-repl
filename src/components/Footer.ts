import { compile } from 'stage1/macro' assert { type: 'macro' };
import { h } from 'stage1/runtime';
import { interpolate, removeNbsp } from '../macros' assert { type: 'macro' };

export type FooterComponent = HTMLElement;

// const meta = compile(`
//   <footer id=foot class="mv3 fss muted tc">
//     © <a href=https://maxmilton.com class="normal muted" rel=noreferrer>Max Milton</a> ・ <a href=https://github.com/maxmilton/ekscss rel=noreferrer>ekscss</a> v${process.env.EKSCSS_VERSION} ・ <a href=https://github.com/maxmilton/ekscss-repl rel=noreferrer>REPL</a> ${process.env.APP_RELEASE} ・ <a href=https://github.com/maxmilton/ekscss-repl/issues rel=noreferrer>report bug</a>
//   </footer>
// `);

const meta = compile(
  // FIXME: This is a convoluted workaround for a bug in the bun macro system,
  // where it crashes when doing string literal template interpolation.
  // https://github.com/oven-sh/bun/issues/3641

  interpolate(
    `
      <footer id=foot class="mv3 fss muted tc">
        ©&nbsp;<a href=https://maxmilton.com class="normal muted" rel=noreferrer>Max Milton</a>&nbsp;・&nbsp;<a href=https://github.com/maxmilton/ekscss rel=noreferrer>ekscss</a>&nbsp;v%%1%% ・&nbsp;<a href=https://github.com/maxmilton/ekscss-repl rel=noreferrer>REPL</a> %%2%% ・&nbsp;<a href=https://github.com/maxmilton/ekscss-repl/issues rel=noreferrer>report bug</a>
      </footer>
    `,
    [process.env.EKSCSS_VERSION!, process.env.APP_RELEASE!],
  ),
);
const view = h<FooterComponent>(removeNbsp(meta.html));

export function Footer(): FooterComponent {
  const root = view;
  return root;
}
