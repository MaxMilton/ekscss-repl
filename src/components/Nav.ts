import h, { HNode } from 'stage0';
import { refs, run } from '../service';
import './Nav.xcss';

type NavComponent = HNode<HTMLDivElement>;

interface RefNodes {
  auto: HTMLInputElement;
  compile: HTMLButtonElement;
  clear: HTMLButtonElement;
}

const view = h`
  <div id=nav class="dfc pv1 ph3">
    <h1 id=logo class=mv0>ekscss REPL</h1>

    <a href=https://maxmilton.github.io/ekscss class=pl4 target=_blank rel=noreferrer>Open ekscss docs</a>

    <label for=auto class="auto label ml-auto mb0 mr3">
      <input id=auto type=checkbox class="checkbox mr2" checked #auto />
      Auto compile on input
    </label>
    <button class="button button-primary mr3" #compile>Compile</button>
    <button class=button #clear>Clear Output</button>
  </div>
`;

export function Nav(): NavComponent {
  const root = view as NavComponent;
  const { auto, compile, clear } = view.collect(root) as RefNodes;

  refs.auto = auto;

  compile.__click = run;

  clear.__click = () => {
    refs.output.setContent('');
    refs.console.clear();
  };

  return root;
}
