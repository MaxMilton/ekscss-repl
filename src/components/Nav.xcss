#nav {
  border-bottom: 1px solid ${x.color.gray5};
}

#logo {
  font-size: 32px;
  font-weight: ${x.app.textWeightHeavy};
}

.logo-icon {
  width: 48px;
}

.github-icon {
  width: 34px;
}

.auto {
  font-weight: ${x.app.textWeightMedium};
  user-select: none;
}
