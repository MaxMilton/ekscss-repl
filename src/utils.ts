const DEFAULT_DEBOUNCE_DELAY_MS = 260;

/**
 * Delay running a function until X ms have passed since its last call.
 */
export function debounce<T extends (
  ...args: unknown[]) => unknown>(
  fn: T,
  delay = DEFAULT_DEBOUNCE_DELAY_MS,
): T {
  let timer: NodeJS.Timeout;

  // @ts-expect-error - Transparent wrapper will not change input function type
  // eslint-disable-next-line func-names
  return function (this: unknown, ...args) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias, unicorn/no-this-assignment
    const context = this;

    clearTimeout(timer);

    timer = setTimeout(() => {
      fn.apply(context, args);
    }, delay);
  };
}
