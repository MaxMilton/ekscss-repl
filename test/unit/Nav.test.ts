import { afterEach, expect, test } from 'bun:test';
import { Nav } from '../../src/components/Nav';
import { cleanup, render } from './utils';

afterEach(cleanup);

test('rendered DOM contains expected elements', () => {
  const rendered = render(Nav());
  expect(rendered.container.firstChild?.nodeName).toBe('DIV');
  expect((rendered.container.firstChild as HTMLDivElement).id).toBe('nav');
  expect(rendered.container.querySelector('h1')?.textContent).toBe('ekscss REPL');
  expect(rendered.container.querySelector('h1')?.id).toBe('logo');
  // TODO: Make test less brittle; we have multiple buttons in Nav.
  expect(rendered.container.querySelector('button')?.textContent).toBe('Compile');

  // TODO: More/better assertions
});

test('rendered DOM matches snapshot', () => {
  const rendered = render(Nav());
  expect(rendered.container.innerHTML).toMatchSnapshot();
});
