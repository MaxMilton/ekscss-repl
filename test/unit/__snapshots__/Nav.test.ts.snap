// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`rendered DOM matches snapshot 1`] = `"<div id="nav" class="dfc pv1 ph3"><h1 id="logo" class="mv0">ekscss REPL</h1><a href="https://ekscss.js.org" class="pl4" target="_blank" rel="noreferrer">Docs</a><div class="df ml-auto"><input id="auto" type="checkbox" class="checkbox" checked=""><label for="auto" class="label">Auto compile on input</label></div><button class="button button-primary mh3">Compile</button><button class="button">Clear Output</button></div>"`;
